### PR TITLE
Update name validation for azurerm_storage_share_directory

### PR DESCRIPTION
**Pull request by neil-yechenwei**
fixes https://github.com/terraform-providers/terraform-provider-azurerm/issues/8341
fixes https://github.com/terraform-providers/terraform-provider-azurerm/issues/8168

--- PASS: TestAccAzureRMStorageShareDirectory_basic (210.22s)
--- PASS: TestAccAzureRMStorageShareDirectory_uppercase (211.34s)
--- PASS: TestAccAzureRMStorageShareDirectory_complete (220.86s)
--- PASS: TestAccAzureRMStorageShareDirectory_requiresImport (227.83s)
--- PASS: TestAccAzureRMStorageShareDirectory_update (269.48s)
--- PASS: TestAccAzureRMStorageShareDirectory_nested (317.45s)

### CODE REVIEW

**Comment by manicminer** [comment]
@neil-yechenwei Thanks for this update. Hope you don't mind, I've pushed some extra changes to support single character directory names, and ensure each directory component conforms. Also, in my testing, double dashes in filenames seemed to work so I removed that check.

**Review by WodansSon — CHANGES_REQUESTED** [review]
@neil-yechenwei Thanks for the PR! It mostly LGTM, however I did leave one very minor comment. Thanks again for the PR 🚀

File: azurerm/helpers/validate/storage.go [hunk 1]
```diff
@@ -10,15 +10,20 @@ func StorageShareDirectoryName(v interface{}, k string) (warnings []string, erro
...
-	// The name cannot contain two consecutive hyphens.
-	if strings.Contains(value, "--") {
-		errors = append(errors, fmt.Errorf("%s cannot contain two concecutive hyphens", k))
+	// The name must begin and end with a letter or a number.
+	start := regexp.MustCompile(`^[a-zA-Z]`)
+	end := regexp.MustCompile(`[a-zA-Z0-9]$`)
+	parts := strings.Split(value, "/")
+	for _, p := range parts {
+		if !start.MatchString(p) || !end.MatchString(p) {
+			errors = append(errors, fmt.Errorf("%s must start and end with a letter and end only with a number or letter", k))
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/helpers/validate/storage.go]: updated

**Comment by neil-yechenwei** [comment]
@manicminer , thanks for your commit, but after tested, I found the name can start with number and hyphen.
image
image
image
image

**Comment by neil-yechenwei** [comment]
@WodansSon , Thanks for your comment. I've updated code according by the newest name validation rules.

**Review by WodansSon — APPROVED** [review]
@neil-yechenwei Thanks for pushing those updates, LGTM! 🚀

**Comment by WodansSon** [comment]
image

**Pull request merged**